packages/@uppy/image-editor/src/ImageEditor.tsx

```diff
@@ -1,14 +1,74 @@
-import { UIPlugin } from '@uppy/core'
+import { UIPlugin, type Uppy } from '@uppy/core'
+import type Cropper from 'cropperjs'
 import { h } from 'preact'
 
-import Editor from './Editor.jsx'
+import type { Meta, Body, UppyFile } from '@uppy/utils/lib/UppyFile'
+import Editor from './Editor.tsx' // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore We don't want TS to generate types for the package.json
 import packageJson from '../package.json'
-import locale from './locale.js'
+import locale from './locale.ts'
 
-export default class ImageEditor extends UIPlugin {
+declare global {
+  namespace preact {
+    interface Component {
+      // This is a workaround for https://github.com/preactjs/preact/issues/1206
+      refs: Record<string, any>
+    }
+  }
+}
+
+type ThumbnailGeneratedCallback<M extends Meta, B extends Body> = (
+  file: UppyFile<M, B>,
+  preview: string,
+) => void
+type GenericCallback<M extends Meta, B extends Body> = (
+  file: UppyFile<M, B>,
+) => void
+declare module '@uppy/core' {
+  export interface UppyEventMap<M extends Meta, B extends Body> {
+    'thumbnail:request': GenericCallback<M, B>
+    'thumbnail:generated': ThumbnailGeneratedCallback<M, B>
+    'file-editor:complete': GenericCallback<M, B>
+    'file-editor:start': GenericCallback<M, B>
+    'file-editor:cancel': GenericCallback<M, B>
+  }
+}
+
+export type Opts = {
+  id?: string
+  target?: string | HTMLElement
+  quality?: number
+  cropperOptions?: Cropper.Options & {
+    croppedCanvasOptions?: Cropper.GetCroppedCanvasOptions
+  }
+  actions?: {
+    revert?: boolean
+    rotate?: boolean
+    granularRotate?: boolean
+    flip?: boolean
+    zoomIn?: boolean
+    zoomOut?: boolean
+    cropSquare?: boolean
+    cropWidescreen?: boolean
+    cropWidescreenVertical?: boolean
+  }
+}
+
+type PluginState<M extends Meta, B extends Body> = {
+  currentImage: UppyFile<M, B> | null
+}
+
+export default class ImageEditor<
+  M extends Meta,
+  B extends Body,
+> extends UIPlugin<Opts, M, B, PluginState<M, B>> {
   static VERSION = packageJson.version
 
-  constructor (uppy, opts) {
+  cropper: Cropper
+
+  opts: Required<Opts>
```

Comment:
Note that in other plugin, I define keys in the `Opts` type as non-optional, and I ask for a `Partial<Opts>` in the constructor. IMO that makes more sense, so you don't have to override this here, and use the one defined in `BasePlugin` instead